File: OJ/leet698/leet698.py
from typing import List

cnt = 0
def gen(L, target):
    def _gen(L, target):
        global cnt
        cnt = cnt + 1
        if not L or target < 0: return []
        if target == 0: return [[]]
        ans = []
        if len(L) > 1 and L[0] + L[-1] <= target:
            for Si in _gen(L[1:], target-L[0]):
                ans.append( [L[0]] + Si )
        return ans + _gen(L[1:], target)
    L = sorted(L, reverse=True)
    return _gen(L, target)


# def check(S: set, L: List[int], ans: List[List]):
#     for si in S: pass
        

# def f(L: List[int], target):
#     S = set(gen(L, target))
#     for si in S:
#         L2 = L.copy()
#         for x in si: L2.remove(x)
#         check(S, L2)


# class Solution:
#     def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
#         if k == 1: return True
#         if sum(nums) % k: return False
#         target = sum(nums) / k 
#         nums.sort(reverse=True)

#         return f(nums, target)

# print(Solution().canPartitionKSubsets( [4, 3, 2, 3, 5, 2, 1] , 4 ))

# print(_gen([4, 3, 2, 3, 5, 2, 1], 5))
print(gen([4, 3, 2, 3, 5, 2, 1, 1, 1, 1, 1, 1], 5))
print(f"cnt = {cnt}")

File: OJ/leet698/leet698.cpp
/*
leet698


*/


#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <tuple>
#include <cassert>
#include <utility>
#include <numeric>
using namespace std;

class Solution {
public:
    int target;
    bool search(vector<int>& B, vector<int>& nums) {
        if (nums.empty()) return true;

        auto v = nums.back(); nums.pop_back();
        for (auto& bucket: B) {
            if (bucket + v <= target) {
                bucket += v;
                if (search(B, nums)) return true;
                bucket -= v;
            }
            if (bucket == 0) break;
        }
        nums.push_back(v);
        return false;
    }
    bool canPartitionKSubsets(const vector<int>& vs, int k) {
        // 4ms
        vector<int> nums = vs;
        if (k == 1) return true;
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (sum % k) return false;

        target = sum / k;

        sort(nums.begin(), nums.end());
        if (nums.back() > target) return false;

        vector<int> B(k); // buckets

        return search(B, nums);
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << boolalpha << sol.canPartitionKSubsets({4, 3, 2, 3, 5, 2, 1}, 4);
    // multiset<int> S = {2, 5, 4, 3, 3, 1, 1, 1};
    // for (auto si: S) cout << si << ", "; cout << endl;

    // S.erase(1);
    // for (auto si: S) cout << si << ", "; cout << endl;

    return 0;
}
